### PR TITLE
Feature/dcc 4672 drug fields

### PR DESCRIPTION
**Pull request by andricDu**
Enforcing allowable fields for drug output. 

Did a little bit of refactor to move the code towards Streams so it is more inline with the rest of the code base. 

I see lots of improvements to be made in general but want to limit the number of moving pieces at the moment. Obvious offender to fix in the future is the writer.

### CODE REVIEW

**Comment by andricDu** [comment]
Field counts compared and matched against historic Drug import (13 fields)

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 1]
```diff
@@ -37,17 +41,46 @@
...
+   * Constants
+   */
   private final static ObjectMapper MAPPER = new ObjectMapper();
-  private final static List<String> GENE_FIELDS_FOR_REMOVE = newArrayList("chembl", "description", "gene_name", "name");
+  private final static List<String> GENE_FIELDS_FOR_REMOVE = newArrayList(
+      "chembl",
+      "description",
+      "gene_name",
+      "name");
+  private final static List<String> ALLOWED_DRUG_FIELDS = newArrayList(
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: Please use Immutable data structures for constants.

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 2]
```diff
@@ -37,17 +41,46 @@
...
-
 @Slf4j
 public class DrugImporter implements SourceImporter {
 
+  /**
+   * Constants
+   */
   private final static ObjectMapper MAPPER = new ObjectMapper();
-  private final static List<String> GENE_FIELDS_FOR_REMOVE = newArrayList("chembl", "description", "gene_name", "name");
+  private final static List<String> GENE_FIELDS_FOR_REMOVE = newArrayList(
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: Please use Immutable data structures for constants.

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 3]
```diff
@@ -64,18 +97,18 @@ public void execute() {
...
-  private List<ObjectNode> readAndJoin(List<ObjectNode> drugs) {
-    return joinTrials(
-        joinGenes(
-            expandImageUrls(
-                denormalizeAtcCodes(
-                    cleanSynonyms(drugs)))));
+    geneMap = new GeneReader(mongoUri).getGeneMap();
+    trialMap = new TrialsReader().getTrialsMap();
+
+    writeDrugs(
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:103]: Nice!

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 4]
```diff
@@ -64,18 +97,18 @@ public void execute() {
...
+        drugs.stream()
+            .map(this::expandImageUrls)
+            .map(this::joinGenes)
+            .map(this::joinTrials)
+            .map(this::denormalizeAtcCodes)
+            .map(this::cleanSynonyms)
+            .map(this::cleanDrug)
+            .collect(toList()));
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `toImmutableList()`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 5]
```diff
@@ -37,17 +41,46 @@
...
+  /**
+   * Dependencies
+   */
   @NonNull
   private final MongoClientURI mongoUri;
 
+  /**
+   * State
+   */
+  private Map<String, ObjectNode> geneMap;
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:81]: Boo state!

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 6]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-          } else {
-            log.warn("Gene missing on join: {}", geneName.asText());
-          }
+  private ObjectNode joinGenes(ObjectNode drug) {
+    val drugGenes = drug.get("genes");
+    val geneArray = MAPPER.createArrayNode();
+
+    if (drugGenes.isArray()) {
+      for (JsonNode geneName : drugGenes) {
+        if (geneMap.containsKey(geneName.asText())) {
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: Double lookup. Just use `get`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java (hunk 2 above)
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `EXCLUDED_GENE_FIELDS`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java (hunk 1 above)
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `INCLUDED_DRUG_FIELDS`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 7]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
    * Creates nodes that contain URLs for the small and large versions of the molecule image
    */
-  private List<ObjectNode> expandImageUrls(List<ObjectNode> drugs) {
-    log.info("Creating image urls");
-
-    drugs.forEach(drug -> {
-      String imageUrl = drug.get("image_url").asText();
-      String largeImageUrl = imageUrl.replace(".png", "-large.png");
+  private ObjectNode expandImageUrls(ObjectNode drug) {
+    String imageUrl = drug.get("image_url").asText();
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `val`s

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 8]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-            trialsArray.add(trialsMap.get(trialCode.asText()));
-          } else {
-            log.warn("Trail missing on join: {}", trialCode.asText());
-          }
+  private ObjectNode joinTrials(ObjectNode drug) {
+    val drugTrials = drug.get("trials");
+    val trialsArray = MAPPER.createArrayNode();
+
+    if (drugTrials.isArray()) {
+      for (JsonNode trialCode : drugTrials) {
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `val`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 9]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-    
-    return drugs;
+  private ObjectNode cleanSynonyms(ObjectNode drug) {
+    val synonyms = (ArrayNode) drug.get("synonyms");
+    val cleaned = MAPPER.createArrayNode();
+
+    if (synonyms != null) {
+      synonyms.forEach(entry -> {
+        if (!entry.asText().equalsIgnoreCase(drug.get("name").asText())) {
+          cleaned.add(entry);
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:193]: We should create a `toArrayNode` collector. Then we could `filter`, `map` and `collect` here.

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 10]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-          }
-        });
-        drug.remove("synonyms");
-        drug.put("synonyms", cleaned);
-      }
-    });
-    
-    return drugs;
+  private ObjectNode cleanSynonyms(ObjectNode drug) {
+    val synonyms = (ArrayNode) drug.get("synonyms");
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:187]: This need to be casted?

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 11]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-          }
+  private ObjectNode joinGenes(ObjectNode drug) {
+    val drugGenes = drug.get("genes");
+    val geneArray = MAPPER.createArrayNode();
+
+    if (drugGenes.isArray()) {
+      for (JsonNode geneName : drugGenes) {
+        if (geneMap.containsKey(geneName.asText())) {
+          ObjectNode cleanedMap = geneMap.get(geneName.asText()).remove(GENE_FIELDS_FOR_REMOVE);
+          geneArray.add(cleanedMap);
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: And here

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 12]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
+    } else {
+      val atcClasses = (ArrayNode) drug.get("atc_classifications");
+      if (atcClasses != null) {
+        ArrayNode newAtcCodes = MAPPER.createArrayNode();
+        atcClasses.forEach(atcClass -> {
+          ObjectNode newAtcEntry = MAPPER.createObjectNode();
+          newAtcEntry.put("code", atcClass.get("level4").asText());
+          newAtcEntry.put("atc_level5_codes", atcClass.get("level5").asText());
+          newAtcEntry.put("description", atcClass.get("level4_description").asText());
+          newAtcCodes.add(newAtcEntry);
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:229]: And here

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 13]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-              ((ObjectNode) atc).put("atc_level5_codes", code.asText());
-              break;
-            }
+  private ObjectNode denormalizeAtcCodes(ObjectNode drug) {
+    val atcCodes = (ArrayNode) drug.get("atc_codes");
+    val level5 = (ArrayNode) drug.get("atc_level5_codes");
+
+    if (atcCodes != null) {
+      atcCodes.forEach(atc -> {
+        for (JsonNode code : level5) {
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:212]: `val`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 14]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-            geneArray.add(cleanedMap);
-          } else {
-            log.warn("Gene missing on join: {}", geneName.asText());
-          }
+  private ObjectNode joinGenes(ObjectNode drug) {
+    val drugGenes = drug.get("genes");
+    val geneArray = MAPPER.createArrayNode();
+
+    if (drugGenes.isArray()) {
+      for (JsonNode geneName : drugGenes) {
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `val`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 15]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
-          } else {
-            log.warn("Trail missing on join: {}", trialCode.asText());
-          }
+  private ObjectNode joinTrials(ObjectNode drug) {
+    val drugTrials = drug.get("trials");
+    val trialsArray = MAPPER.createArrayNode();
+
+    if (drugTrials.isArray()) {
+      for (JsonNode trialCode : drugTrials) {
+        if (trialMap.containsKey(trialCode.asText())) {
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: Double lookup. just use `get` with `null` check.

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 16]
```diff
@@ -89,137 +122,122 @@ private void writeDrugs(List<ObjectNode> drugs) {
...
+        drug.remove("atc_classifications");
       }
-    });
+    }
+
+    return drug;
+  }
 
-    return drugs;
+  private ObjectNode cleanDrug(ObjectNode drug) {
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java]: `static`

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java (hunk 10 above)
Comment (andricDu) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:187]: I could use `withArray` but I need to be sure that creating the array is safe.

Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:187]: What code relies on it being an array?

File: dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java [hunk 17]
```diff
@@ -107,8 +106,8 @@ public void execute() {
             .map(this::joinTrials)
             .map(this::denormalizeAtcCodes)
             .map(this::cleanSynonyms)
-            .map(this::cleanDrug)
-            .collect(toList()));
+            .map(DrugImporter::cleanDrug)
```
Comment (btiernay) [inline dcc-import-drug/src/main/java/org/icgc/dcc/imports/drug/DrugImporter.java:110]: Ok, I see why you did that now :)

**Pull request merged**